File: problems/008-mounted-directory-baseline/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func pathLess(a, b []uint32) bool {
	limit := len(a)
	if len(b) < limit {
		limit = len(b)
	}
	for i := 0; i < limit; i++ {
		if a[i] != b[i] {
			return a[i] < b[i]
		}
	}
	return len(a) < len(b)
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var mountedCount, outputCount int
	var byteQuota, inodeQuota uint64
	if _, err := fmt.Fscan(in, &mountedCount, &outputCount, &byteQuota, &inodeQuota); err != nil {
		return
	}

	pathCount := mountedCount + outputCount
	paths := make([][]uint32, pathCount)
	var baselineBytes uint64
	for i := 0; i < pathCount; i++ {
		var length int
		fmt.Fscan(in, &length)
		paths[i] = make([]uint32, length)
		for j := range paths[i] {
			fmt.Fscan(in, &paths[i][j])
		}
		if i < mountedCount {
			var size uint64
			fmt.Fscan(in, &size)
			baselineBytes += size
		}
	}

	sort.Slice(paths, func(i, j int) bool { return pathLess(paths[i], paths[j]) })
	directoryCount := uint64(1)
	for i, path := range paths {
		parentLength := len(path) - 1
		alreadyPresent := 0
		if i > 0 {
			limit := len(paths[i-1])
			if len(path) < limit {
				limit = len(path)
			}
			for alreadyPresent < limit && paths[i-1][alreadyPresent] == path[alreadyPresent] {
				alreadyPresent++
			}
			if alreadyPresent > parentLength {
				alreadyPresent = parentLength
			}
		}
		directoryCount += uint64(parentLength - alreadyPresent)
	}

	baselineInodes := directoryCount + uint64(pathCount)
	if baselineBytes <= byteQuota && baselineInodes <= inodeQuota {
		fmt.Fprintln(out, "ACCEPT", baselineBytes, baselineInodes,
			byteQuota-baselineBytes, inodeQuota-baselineInodes)
	} else {
		var missingBytes, missingInodes uint64
		if baselineBytes > byteQuota {
			missingBytes = baselineBytes - byteQuota
		}
		if baselineInodes > inodeQuota {
			missingInodes = baselineInodes - inodeQuota
		}
		fmt.Fprintln(out, "REJECT", baselineBytes, baselineInodes, missingBytes, missingInodes)
	}
}

File: problems/008-mounted-directory-baseline/solutions/javascript/main.js
import * as std from "std";

/** @type {string} */
const input = std.in.readAsString();
let cursor = 0;
/** @returns {string} */
function nextToken() {
  while (cursor < input.length && input.charCodeAt(cursor) <= 32) cursor++;
  const start = cursor;
  while (cursor < input.length && input.charCodeAt(cursor) > 32) cursor++;
  return input.slice(start, cursor);
}
const mountedCount = Number(nextToken());
const outputCount = Number(nextToken());
const byteQuota = BigInt(nextToken());
const inodeQuota = BigInt(nextToken());

const pathCount = mountedCount + outputCount;
const paths = /** @type {number[][]} */ ([]);
let baselineBytes = 0n;
for (let i = 0; i < pathCount; i++) {
  const length = Number(nextToken());
  const path = /** @type {number[]} */ ([]);
  for (let j = 0; j < length; j++) path.push(Number(nextToken()));
  paths.push(path);
  if (i < mountedCount) baselineBytes += BigInt(nextToken());
}

paths.sort((a, b) => {
  const limit = Math.min(a.length, b.length);
  for (let i = 0; i < limit; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
});

let directoryCount = 1n;
for (let i = 0; i < pathCount; i++) {
  const parentLength = paths[i].length - 1;
  let alreadyPresent = 0;
  if (i > 0) {
    const limit = Math.min(paths[i - 1].length, paths[i].length);
    while (
      alreadyPresent < limit &&
      paths[i - 1][alreadyPresent] === paths[i][alreadyPresent]
    ) {
      alreadyPresent++;
    }
    alreadyPresent = Math.min(alreadyPresent, parentLength);
  }
  directoryCount += BigInt(parentLength - alreadyPresent);
}

const baselineInodes = directoryCount + BigInt(pathCount);
const accepted = baselineBytes <= byteQuota && baselineInodes <= inodeQuota;
const missingBytes = baselineBytes > byteQuota ? baselineBytes - byteQuota : 0n;
const missingInodes = baselineInodes > inodeQuota
  ? baselineInodes - inodeQuota
  : 0n;
std.out.puts(
  `${accepted ? "ACCEPT" : "REJECT"} ${baselineBytes} ${baselineInodes} ` +
    `${accepted ? byteQuota - baselineBytes : missingBytes} ` +
    `${accepted ? inodeQuota - baselineInodes : missingInodes}\n`,
);
